cap-primitives/src/std/fs/get_path.rs

```diff
@@ -0,0 +1,35 @@
+//! `get_path` functions similar to the code in libstd.
+
+use std::{fs, os::unix::io::AsRawFd, path::PathBuf};
+
+#[cfg(target_os = "linux")]
+pub(crate) fn get_path(file: &fs::File) -> Option<PathBuf> {
+    let mut p = PathBuf::from("/proc/self/fd");
+    p.push(&file.as_raw_fd().to_string());
+    fs::read_link(p).ok()
+}
+
+#[cfg(target_os = "macos")]
+pub(crate) fn get_path(file: &fs::File) -> Option<PathBuf> {
+    use std::os::unix::ffi::OsStringExt;
+
+    // The use of PATH_MAX is generally not encouraged, but it
+    // is inevitable in this case because macOS defines `fcntl` with
+    // `F_GETPATH` in terms of `MAXPATHLEN`, and there are no
+    // alternatives. If a better method is invented, it should be used
+    // instead.
+    let mut buf = vec![0; libc::PATH_MAX as usize];
+    let n = unsafe { libc::fcntl(file.as_raw_fd(), libc::F_GETPATH, buf.as_ptr()) };
+    if n == -1 {
+        return None;
+    }
+    let l = buf.iter().position(|&c| c == 0).unwrap();
+    buf.truncate(l as usize);
+    buf.shrink_to_fit();
```

Comment:
This is super unfortunate and alas I've not found a better way at solving this on Darwin... In fact, I've got pretty much the exact same solution over in Zig: [zig#5701](https://github.com/ziglang/zig/pull/5701/files#diff-5a679ae5685f723b7d3fec367ffbeab6R3951).